File: src/csharp/pulse/cdm/scalars/SEScalarTemperature.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

public class TemperatureUnit
{
  private TemperatureUnit(string v) { Value = v; }
  protected string Value { get; }

  public new string ToString() { return Value; }

  public static TemperatureUnit C = new TemperatureUnit("degC");
  public static TemperatureUnit F = new TemperatureUnit("degF");
  public static TemperatureUnit K = new TemperatureUnit("K");
  public static TemperatureUnit R = new TemperatureUnit("degR");

  public static TemperatureUnit FromString(string u)
  {
    if (u == TemperatureUnit.C.ToString())
      return TemperatureUnit.C;
    if (u == TemperatureUnit.F.ToString())
      return TemperatureUnit.F;
    if (u == TemperatureUnit.K.ToString())
      return TemperatureUnit.K;
    if (u == TemperatureUnit.R.ToString())
      return TemperatureUnit.R;

    throw new System.ArgumentException(u + " is not a know temperature unit");
  }
}

public class SEScalarTemperature : SEScalar
{

  protected TemperatureUnit unit;

  public SEScalarTemperature() : base()
  {

  }

  public SEScalarTemperature(SEScalarTemperature from)
  {
    Set(from);
  }

  public SEScalarTemperature(double value, TemperatureUnit unit)
  {
    SetValue(value, unit);
  }

  public void Set(SEScalarTemperature from)
  {
    if (from == null)
    {
      Invalidate();
      return;
    }
    SetValue(from.value, from.unit);
  }

  public new void SetValue(double value)
  {
    throw new System.NotSupportedException("You must provide a unit");
  }

  public void SetValue(double value, TemperatureUnit unit)
  {
    this.value = value;
    this.unit = unit;
  }

  public double GetValue(TemperatureUnit unit)
  {
    throw new System.NotImplementedException();
  }

  public TemperatureUnit GetUnit()
  {
    return unit;
  }

  public new string ToString()
  {
    return this.value + "(" + this.unit.ToString() + ")";
  }
}

File: src/cpp/cdm/properties/SEScalarTemperature.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarTemperature.h"

const TemperatureUnit TemperatureUnit::F("degF");
const TemperatureUnit TemperatureUnit::C("degC");
const TemperatureUnit TemperatureUnit::K("K");
const TemperatureUnit TemperatureUnit::R("degR");

bool TemperatureUnit::IsValidUnit(const std::string& unit)
{
  if (F.GetString().compare(unit) == 0)
    return true;
  if (C.GetString().compare(unit) == 0)
    return true;
  if (K.GetString().compare(unit) == 0)
    return true;
  if (R.GetString().compare(unit) == 0)
    return true;
  return false;
}

const TemperatureUnit& TemperatureUnit::GetCompoundUnit(const std::string& unit)
{
  if (F.GetString().compare(unit) == 0)
    return F;
  if (C.GetString().compare(unit) == 0)
    return C;
  if (K.GetString().compare(unit) == 0)
    return K;
  if (R.GetString().compare(unit) == 0)
    return R;
  throw CommonDataModelException(unit + " is not a valid Temperature unit");
}

double SEScalarTemperature::GetValue(const TemperatureUnit& unit) const
{
  if (m_isnan)
    throw CommonDataModelException("Value is NaN");
  if (m_isinf)
    return m_value;
  if (m_unit == &unit)
    return m_value;
  return Convert(m_value, *m_unit, unit);
}

template class SEScalarQuantity<TemperatureUnit>;
